prereise/gather/griddata/hifld/data_process/demand.py

```diff
@@ -0,0 +1,81 @@
+from prereise.gather.griddata.hifld import const
+
+
+def assign_demand_to_buses(substations, branch, plant, bus):
+    """Using data on population by county and ZIP code, assign demand to substations,
+    then to the lowest-voltage bus within each substation.
+    This demand parameter is added inplace as a 'Pd' column to the ``bus`` data frame.
+
+    :param pandas.DataFrame substations: table of substation data.
+    :param pandas.DataFrame branch: table of branch data.
+    :param pandas.DataFrame plant: table of plant data.
+    :param pandas.DataFrame bus: table of bus data.
+    """
+    # Load data
+    data_dir = os.path.join(os.path.dirname(inspect.getfile(hifld)), "data")
+    zip_path = os.path.join(data_dir, "uszips.csv")
+    zip_data = pd.read_csv(zip_path, dtype={"zip": "string"}).set_index("zip")
+    county_data = pd.read_csv(os.path.join(data_dir, "uscounties.csv"))
+
+    # Determine each substation's transmission capacity, then sort for selection
+    filtered_branch = branch.query("SUB_1_ID != SUB_2_ID")
+    from_cap = filtered_branch.groupby("SUB_1_ID").sum()["rateA"]
+    to_cap = filtered_branch.groupby("SUB_2_ID").sum()["rateA"]
+    sub_cap = from_cap.combine(to_cap, lambda x, y: x + y, fill_value=0)
+    # Sort substations by their capacities for later ordered selection
+    sorted_subs = substations.loc[sub_cap.sort_values(ascending=False).index].copy()
+
+    # Determine for each ZIP, how much demand to assign to each load substation
+    # Assume here that generator substations don't have load attached to them
+    filtered_subs = sorted_subs.loc[~sorted_subs.index.isin(plant["sub_id"])]
+    subs_per_zip = filtered_subs.value_counts("ZIP")
+    zip_load_substations = subs_per_zip * const.substation_load_share
+    zip_load_substations = zip_load_substations.round().clip(lower=1)
+    zip_assigned_population = (zip_data["population"] / zip_load_substations).dropna()
+    # Select the N substations per ZIP with greatest transmission capacity
+    load_substations = pd.concat(
+        df.head(int(zip_load_substations[name]))
+        for name, df in filtered_subs.groupby("ZIP")
+    )
+    substations["pop_ZIP"] = load_substations["ZIP"].map(zip_assigned_population)
+
+    # Assign remaining county population to substations with load already,
+    #   plus the most connected substation in any county without a load substation
+    load_subs_from_zips = substations.query("pop_ZIP > 0")
+    load_subs_per_county = load_subs_from_zips.value_counts("COUNTYFIPS")
+    county_pop = county_data.set_index("county_fips")["population"]
+
+    # Select the one substation per missing county with greatest transmission capacity
+    counties_without_load_subs = set(county_pop.index) - set(load_subs_per_county.index)
+    subs_in_counties_without_load_subs = sorted_subs.loc[
+        sorted_subs["COUNTYFIPS"].isin(counties_without_load_subs)
+    ]
+    added_load_subs = pd.concat(
+        df.head(1)
```

Comment:
I was thinking whether it is worth doing a similar N distribution instead of picking the one with largest transmission capacity.